extensions-core/druid-kerberos/src/main/java/io/druid/kerberos/KerberosHttpClient.java

```diff
@@ -0,0 +1,141 @@
+import org.ietf.jgss.GSSManager;
+import org.ietf.jgss.GSSName;
+import org.ietf.jgss.Oid;
+import org.jboss.netty.handler.codec.http.HttpHeaders;
+import org.joda.time.Duration;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+
+public class KerberosHttpClient extends AbstractHttpClient
+{
+  private static final Logger log = new Logger(KerberosHttpClient.class);
+
+  private final HttpClient delegate;
+  private final DruidKerberosConfig config;
+  private static final Base64 base64codec = new Base64(0);
+
+
+  public KerberosHttpClient(HttpClient delegate, DruidKerberosConfig config)
+  {
+    this.delegate = delegate;
+    this.config = config;
+  }
+
+  @Override
+  public <Intermediate, Final> ListenableFuture<Final> go(
+    Request request, HttpResponseHandler<Intermediate, Final> httpResponseHandler, Duration duration
+  )
+  {
+    try {
+      final String host = request.getUrl().getHost();
+      authenticateIfRequired();
+      UserGroupInformation currentUser = UserGroupInformation.getCurrentUser();
+      String challenge = currentUser.doAs(new PrivilegedExceptionAction<String>()
+      {
+        @Override
+        public String run() throws Exception
+        {
+          return kerberosChallenge(host);
+        }
+      });
+      request.setHeader(HttpHeaders.Names.AUTHORIZATION, "Negotiate " + challenge);
+    }
+    catch (Throwable e) {
+      Throwables.propagate(e);
+    }
+    return delegate.go(request, httpResponseHandler, duration);
+  }
+
+
+  public void authenticateIfRequired()
+    throws IOException
+  {
+    String principal = config.getPrincipal();
+    String keytab = config.getKeytab();
+    if (!Strings.isNullOrEmpty(principal) && !Strings.isNullOrEmpty(keytab)) {
+      Configuration conf = new Configuration();
+      conf.set(CommonConfigurationKeysPublic.HADOOP_SECURITY_AUTHENTICATION, "kerberos");
+      UserGroupInformation.setConfiguration(conf);
+      if (UserGroupInformation.isSecurityEnabled()) {
+        try {
+          if (UserGroupInformation.getCurrentUser().hasKerberosCredentials() == false
+              || !UserGroupInformation.getCurrentUser().getUserName().equals(principal)) {
+            log.info("trying to authenticate user [%s] with keytab [%s]", principal, keytab);
+            UserGroupInformation.loginUserFromKeytab(principal, keytab);
+          }
+        }
+        catch (IOException e) {
+          throw new ISE(e, "Failed to authenticate user principal [%s] with keytab [%s]", principal, keytab);
+        }
+      }
+    }
+  }
+
+  /**
+   * This method always needs to be called within a doAs block so that the client's TGT credentials
+   * can be read from the Subject.
+   *
+   * @return Kerberos Challenge String
+   *
+   * @throws Exception
+   */
+
+  public static String kerberosChallenge(String server) throws GSSException
+  {
+    // This Oid for Kerberos GSS-API mechanism.
+    Oid mechOid = new Oid("1.2.840.113554.1.2.2");
```

Comment:
looking at Hadoop project looks like this is JVM dependent https://github.com/apache/hadoop/blob/916140604ffef59466ba30832478311d3e6249bd/hadoop-common-project/hadoop-auth/src/main/java/org/apache/hadoop/security/authentication/util/KerberosUtil.java#L54-L54